In [53]:
import numpy as np 
import pandas as pd

In [54]:
prices = pd.read_csv("../data/raw/train.csv")
print("Shape of prices: {}".format(prices.shape))

Shape of prices: (1460, 81)


## Exploring data

Which columns have NA values?

In [55]:
print("Which columns have NA values?")
nas = prices.isnull().any()
print(nas[nas == True])

Which columns have NA values?
LotFrontage     True
Alley           True
MasVnrType      True
MasVnrArea      True
BsmtQual        True
BsmtCond        True
BsmtExposure    True
BsmtFinType1    True
BsmtFinType2    True
Electrical      True
FireplaceQu     True
GarageType      True
GarageYrBlt     True
GarageFinish    True
GarageQual      True
GarageCond      True
PoolQC          True
Fence           True
MiscFeature     True
dtype: bool


Classification of features into Nominal/Ordinal/Quantitative has been done in /reports/Data-classification.ods

In [56]:
nominals = ['MSSubClass', 'MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
            'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
            'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
            'Foundation', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 'CentralAir', 'Electrical',
            'Functional', 'GarageType', 'GarageFinish', 'PavedDrive', 'Fence', 'MiscFeature', 'SaleType',
            'SaleCondition']

ordinals = ['OverallQual', 'OverallCond', 'ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond', 'BsmtExposure',
           'HeatingQC', 'KitchenQual', 'FireplaceQu', 'GarageQual','GarageCond', 'PoolQC']

quant_ratio = ['YearBuilt', 'YearRemodAdd', 'GarageYrBlt', 'MoSold', 'YrSold']

quant_int = ['LotFrontage', 'LotArea', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
           'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
           'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 
           'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageCars',
           'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 
           'ScreenPorch', 'PoolArea', 'MiscVal', 'SalePrice']

In [69]:
prices.describe(include=[np.number])

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726   
std       1.112799    30.202904     20.645407   181.066207   456.098091   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000   

           ...         WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  \
count      ...        1460.000000  1460.000000    1460.000000  1460.000000   
mean       ...          94.244521    46.660274      21.954110     3.409589   
std        ...         125.338794    66.256028      61.119149    29.317331   
min        ...           0.000000     0.000000       0.000000     0.000000   
25%        ...           0.000000     0.000000       0.000000     0.000000   
50%        ...           0.000000    25.000000       0.000000     0.000000   
75%        ...         168.000000    68.000000       0.000000     0.000000   
max        ...         857.000000   547.000000     552.000000   508.000000   

       ScreenPorch     PoolArea       MiscVal       MoSold       YrSold  \
count  1460.000000  1460.000000   1460.000000  1460.000000  1460.000000   
mean     15.060959     2.758904     43.489041     6.321918  2007.815753   
std      55.757415    40.177307    496.123024     2.703626     1.328095   
min       0.000000     0.000000      0.000000     1.000000  2006.000000   
25%       0.000000     0.000000      0.000000     5.000000  2007.000000   
50%       0.000000     0.000000      0.000000     6.000000  2008.000000   
75%       0.000000     0.000000      0.000000     8.000000  2009.000000   
max     480.000000   738.000000  15500.000000    12.000000  2010.000000   

           SalePrice  
count    1460.000000  
mean   180921.195890  
std     79442.502883  
min     34900.000000  
25%    129975.000000  
50%    163000.000000  
75%    214000.000000  
max    755000.000000  

[8 rows x 38 columns]


Let's check the nominal features to get an idea of the relative frequency of each value.

In [58]:
for nom in nominals:
    
    print(nom+" Has null values? {}\n".format(prices[nom].isnull().any()))
    print(nom+": Different classes \n{}\n".format(np.unique(prices[~prices[nom].isnull()][nom])))
    print(nom+": Occurrences per class\n{}".format(prices[nom].value_counts()))
    print("\n--------------------------------------\n")


MSSubClass Has null values? False

MSSubClass: Different classes 
[ 20  30  40  45  50  60  70  75  80  85  90 120 160 180 190]

MSSubClass: Occurrences per class
20     536
60     299
50     144
120     87
30      69
160     63
70      60
80      58
90      52
190     30
85      20
75      16
45      12
180     10
40       4
Name: MSSubClass, dtype: int64

--------------------------------------

MSZoning Has null values? False

MSZoning: Different classes 
['C (all)' 'FV' 'RH' 'RL' 'RM']

MSZoning: Occurrences per class
RL         1151
RM          218
FV           65
RH           16
C (all)      10
Name: MSZoning, dtype: int64

--------------------------------------

Street Has null values? False

Street: Different classes 
['Grvl' 'Pave']

Street: Occurrences per class
Pave    1454
Grvl       6
Name: Street, dtype: int64

--------------------------------------

Alley Has null values? True

Alley: Different classes 
['Grvl' 'Pave']

Alley: Occurrences per class
Grvl    50
Pave    41
N

We see that for most features one or two values occur significantly more often than others.  

Next let's do the same check for the ordinal features.

In [59]:
for nom in ordinals:
    
    print(nom+" Has null values? {}\n".format(prices[nom].isnull().any()))
    print(nom+": Different classes \n{}\n".format(np.unique(prices[~prices[nom].isnull()][nom])))
    print(nom+": Occurrences per class\n{}".format(prices[nom].value_counts()))
    print("\n--------------------------------------\n")


OverallQual Has null values? False

OverallQual: Different classes 
[ 1  2  3  4  5  6  7  8  9 10]

OverallQual: Occurrences per class
5     397
6     374
7     319
8     168
4     116
9      43
3      20
10     18
2       3
1       2
Name: OverallQual, dtype: int64

--------------------------------------

OverallCond Has null values? False

OverallCond: Different classes 
[1 2 3 4 5 6 7 8 9]

OverallCond: Occurrences per class
5    821
6    252
7    205
8     72
4     57
3     25
9     22
2      5
1      1
Name: OverallCond, dtype: int64

--------------------------------------

ExterQual Has null values? False

ExterQual: Different classes 
['Ex' 'Fa' 'Gd' 'TA']

ExterQual: Occurrences per class
TA    906
Gd    488
Ex     52
Fa     14
Name: ExterQual, dtype: int64

--------------------------------------

ExterCond Has null values? False

ExterCond: Different classes 
['Ex' 'Fa' 'Gd' 'Po' 'TA']

ExterCond: Occurrences per class
TA    1282
Gd     146
Fa      28
Ex       3
Po       1
Na

While the features OverallQual and OverallCond have numeric values, the other ordinal features have string values. The set of string values is almost the same over the ordinal features. Those values can be uniformized and then transformed into numeric values to allow an ordering between them.

## Sanity checking data

Check for duplicates and partial duplicates.

In [60]:
dups = prices.duplicated()
print(dups[dups == True])

Series([], dtype: bool)


No duplicate rows in the dataset

In [61]:
# which columns could help most to identify partial duplicates?
# Neighborhood - duplicates should be in the same neighborhood
# MoSold / YrSold - sold at the same time
# YearBuilt - the year it was built
# SalePrice - sold for the same price

dupsgrouped = prices[prices.duplicated(['Neighborhood', 'YrSold', 'MoSold', 'YearBuilt', 'SalePrice'], keep=False)].groupby(
    ['Neighborhood', 'YrSold', 'MoSold', 'YearBuilt', 'SalePrice'])
for name, dup in dupsgrouped:
    for col in dup:
        print(col)
        print(dup[col][:])
        print('\n')
    print('------------')

Id
145      146
1304    1305
Name: Id, dtype: int64


MSSubClass
145     160
1304    160
Name: MSSubClass, dtype: int64


MSZoning
145     RM
1304    RM
Name: MSZoning, dtype: object


LotFrontage
145     24.0
1304    32.0
Name: LotFrontage, dtype: float64


LotArea
145     2522
1304    3363
Name: LotArea, dtype: int64


Street
145     Pave
1304    Pave
Name: Street, dtype: object


Alley
145     NaN
1304    NaN
Name: Alley, dtype: object


LotShape
145     Reg
1304    Reg
Name: LotShape, dtype: object


LandContour
145     Lvl
1304    Lvl
Name: LandContour, dtype: object


Utilities
145     AllPub
1304    AllPub
Name: Utilities, dtype: object


LotConfig
145     Inside
1304    Inside
Name: LotConfig, dtype: object


LandSlope
145     Gtl
1304    Gtl
Name: LandSlope, dtype: object


Neighborhood
145     Edwards
1304    Edwards
Name: Neighborhood, dtype: object


Condition1
145     Norm
1304    Norm
Name: Condition1, dtype: object


Condition2
145     Norm
1304    Norm
Name: Condition2,

Only potential candidates for partial duplicate are row 145 (Id 146) and 1304 (Id 1305) - they are however sufficiently different on the LotArea to consider we are dealing with two different house sales. 

The total basement surface should correspond to the sum of BsmtFin1SF, BsmtFin2SF and BsmtUnfSF. Is this true for all records?

In [62]:
#TotalBsmtSF = BsmtFin1SF + BsmtFin2SF + BsmtUnfSF -YES
bools = prices['TotalBsmtSF'] == prices['BsmtFinSF1'] + prices['BsmtFinSF2'] + prices['BsmtUnfSF']
prices[~bools.values]


Empty DataFrame
Columns: [Id, MSSubClass, MSZoning, LotFrontage, LotArea, Street, Alley, LotShape, LandContour, Utilities, LotConfig, LandSlope, Neighborhood, Condition1, Condition2, BldgType, HouseStyle, OverallQual, OverallCond, YearBuilt, YearRemodAdd, RoofStyle, RoofMatl, Exterior1st, Exterior2nd, MasVnrType, MasVnrArea, ExterQual, ExterCond, Foundation, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinSF1, BsmtFinType2, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, Heating, HeatingQC, CentralAir, Electrical, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, KitchenQual, TotRmsAbvGrd, Functional, Fireplaces, FireplaceQu, GarageType, GarageYrBlt, GarageFinish, GarageCars, GarageArea, GarageQual, GarageCond, PavedDrive, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, PoolQC, Fence, MiscFeature, MiscVal, MoSold, YrSold, SaleType, SaleCondition, SalePrice]
Index: []

[0 rows x 81 columns]

No records found where this isn't verified.
To reduce the number of variables we can add BmsFinSF1 and BsmtFinSF2 to obtain the total BsmtFinSF and eliminate TotalBsmtSF, BsmtFinSF1 and BsmtFinSF2.

Now check the different values of the quantitative ratio features. Those features describe a year or a month.

In [63]:
for nom in quant_ratio:
    
    print(nom+" Has null values? {}\n".format(prices[nom].isnull().any()))
    print(nom+": Different classes \n{}\n".format(np.unique(prices[~prices[nom].isnull()][nom])))
    print(nom+": Occurrences per class\n{}".format(prices[nom].value_counts()))
    print("\n--------------------------------------\n")




YearBuilt Has null values? False

YearBuilt: Different classes 
[1872 1875 1880 1882 1885 1890 1892 1893 1898 1900 1904 1905 1906 1908 1910
 1911 1912 1913 1914 1915 1916 1917 1918 1919 1920 1921 1922 1923 1924 1925
 1926 1927 1928 1929 1930 1931 1932 1934 1935 1936 1937 1938 1939 1940 1941
 1942 1945 1946 1947 1948 1949 1950 1951 1952 1953 1954 1955 1956 1957 1958
 1959 1960 1961 1962 1963 1964 1965 1966 1967 1968 1969 1970 1971 1972 1973
 1974 1975 1976 1977 1978 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988
 1989 1990 1991 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003
 2004 2005 2006 2007 2008 2009 2010]

YearBuilt: Occurrences per class
2006    67
2005    64
2004    54
2007    49
2003    45
1976    33
1977    32
1920    30
1959    26
1999    25
1998    25
1958    24
1965    24
1970    24
1954    24
2000    24
2002    23
2008    23
1972    23
1968    22
1971    22
1950    20
2001    20
1957    20
1962    19
1994    19
1966    18
2009    18
1995    18
1940    18
   

The values to describe years and months seem correct (months range from 1 - 12, years from 1872 - 2010)

One would assume that some basic constraints should hold:  
- the house is built before it is sold (YearBuilt < YrSold)
- the house is built before it is remodelled (YearBuilt < YearRemodAdd)
- the house is built before or in the same year the garage is built


In [64]:
bools = prices['YearBuilt'] <= prices['YrSold']
print(prices[~bools][['YearBuilt','YrSold']])

bools = prices['YearBuilt'] <= prices['YearRemodAdd'].fillna(3000)
print(prices[~bools][['YearBuilt','YearRemodAdd']])

bools = prices['YearBuilt'] <= prices['GarageYrBlt'].fillna(3000)
print(prices[~bools][['YearBuilt','GarageYrBlt']])

Empty DataFrame
Columns: [YearBuilt, YrSold]
Index: []
Empty DataFrame
Columns: [YearBuilt, YearRemodAdd]
Index: []
      YearBuilt  GarageYrBlt
29         1927       1920.0
93         1910       1900.0
324        1967       1961.0
600        2005       2003.0
736        1950       1949.0
1103       1959       1954.0
1376       1930       1925.0
1414       1923       1922.0
1418       1963       1962.0


The first two constraints hold, however there are 9 cases where the garage was built earlier than the house.


In [65]:
nogarage = len(prices[prices['GarageYrBlt'].isnull()])
sameyear = len(prices[prices['YearBuilt'] == prices['GarageYrBlt'].fillna(3000)])
later = len(prices[prices['YearBuilt'] < prices['GarageYrBlt'].fillna(prices['YearBuilt'])])                
earlier = len(prices[prices['YearBuilt'] > prices['GarageYrBlt'].fillna(prices['YearBuilt'])])
               
print("Garage was never built : {}".format(nogarage))
print("Gargae was built same year as house : {}".format(sameyear))
print("Gargae was built later than house : {}".format(later))
print("Gargae was built earlier than house : {}".format(earlier))               

Garage was never built : 81
Gargae was built same year as house : 1089
Gargae was built later than house : 281
Gargae was built earlier than house : 9


We see that for the houses that have a garage (which is the majority) the garage was built the same year as the house for 79%. It seems unusual to have a garage built earlier than the house, those 9 records should be corrected and the value of GarageYrBlt should be defaulted to YearBuilt.